### DISCUSSION TITLE
conn.close() but MS Access .laccdb (lock) file stays open

### DISCUSSION THREAD

**Discussion by ScotterMonk** [mkleehammer/pyodbc#1280] [Q&A]
# Environment:
- Python 3.11 
- pyodbc latest
- OS: Windows 11
- DB: MS Access 2019
- driver: ODBC

# My issue:
Open database, read & write <-- all works fine.
At end conn.close()
.Laccdb file (MS Access lock file) stays visible in folder next to the .accdb file.
It's not a "slow" thing. I refresh, wait, etc. and that .Laccdb file is still there.
I *can* delete it manually but would like this to work the way it is supposed to.
Until I can figure out an elegant fix, I'll have my program delete the file.

**Comment by gordthompson** [comment]
Does the .laccdb file remain if you simply open the connection, wait a couple of seconds, and then close the connection?

```python
import time

import pyodbc

connection_string = "Driver={Microsoft Access Driver (*.mdb, *.accdb)};DBQ=…"
cnxn = pyodbc.connect(connection_string)
sleep_time = 3
print(f"Connected. Pausing for {sleep_time} seconds.")
time.sleep(sleep_time)
print("Closing connection.")
cnxn.close()
```

**Comment by v-chojas** [comment]
It may be a driver bug, could you post an ODBC trace? The driver is supposed to cleanup upon connection close but perhaps it isn't doing that.

**Comment by gordthompson** [comment]
Also, when you say "At end conn.close()" does it mean that your script terminates immediately after the `.close()`?

**Comment by ScotterMonk** [comment]
> Also, when you say "At end conn.close()" does it mean that your script terminates immediately after the `.close()`?

Yes and I'm trying your suggestion to put a short sleep after the close.

**Comment by gordthompson** [comment]
FWIW, I am unable to reproduce your issue with ACE for Access 2013 and an .accdb on my local hard drive. The .laccdb file is removed within one second of calling `.close()` on the connection, even if the script continues to run. Is your .accdb on a network share, by chance?

**Comment by ScotterMonk** [comment]
Aha! Dropbox folder.
BTW still have not had time to test with new pause-code you recommended.

Sincerely,
Scott Swain
https://OceanMedia.net

Sent with [Proton Mail](https://proton.me/) secure email.

------- Original Message -------

**Reply by gordthompson** [reply]
I'm also unable to reproduce the issue with the .accdb file in a Dropbox folder.

```python
import os
import time

import msaccessdb  # needs: pip install msaccessdb
import pyodbc

file_stub = r"C:\Users\Gord\Dropbox\__tmp\laccdb_test."
accdb_spec = file_stub + "accdb"
laccdb_spec = file_stub + "laccdb"

if not os.path.isfile(accdb_spec):
    msaccessdb.create(accdb_spec)

cnxn = pyodbc.connect(
    "Driver={Microsoft Access Driver (*.mdb, *.accdb)};"
    f"DBQ={accdb_spec}"
)
print("Connected. Pausing for 3 seconds.")
time.sleep(3)
elapsed = 0
interval = 5  # seconds
limit = 60
cnxn.close()
print("Connection closed ...")
while elapsed < limit:
    time.sleep(interval)
    elapsed += interval
    laccdb_exists = os.path.isfile(laccdb_spec)
    print(f"... after {elapsed} seconds, laccdb file {'still exists' if laccdb_exists else 'is gone'}")
    if not laccdb_exists:
        break
```

Console output:

```
Connected. Pausing for 3 seconds.
Connection closed ...
... after 5 seconds, laccdb file is gone
```

although it is still possible that when the connection is closed the Dropbox sync may interfere with dropping the .laccdb file.
